File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
    string = input()
    if string == '.':
        break
    else:
        string = list(string)
        l = []
        for i in range(len(string)):
            if string[i] in ['(', ')', '[', ']']:
                if len(l) == 0:
                    l.append(string[i])
                else:
                    if l[-1] == '(' and string[i] == ')':
                        l.pop()
                    elif l[-1] == '[' and string[i] == ']':
                        l.pop()
                    else:
                        l.append(string[i])
        if len(l) == 0:
            print("yes")
        else:
            print("no")

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <bits/stdc++.h>
using namespace std;

string str;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    while (true) {
        stack<char> s;
        bool check = true;

        getline(cin, str);
        if (str == ".") break;
        for (int i=0; i<str.size(); i++) {
            if (str[i] == '(') {
                s.push('(');
            } else if (str[i] == '[') {
                s.push('[');
            } else if (str[i] == ')') {
                if (s.empty()) {
                    check = false;
                    break;
                } else {
                    if (s.top() == '(') {
                        s.pop();
                    } else {
                       check = false;
                        break; 
                    }
                }
            } else if (str[i] == ']') {
                if (s.empty()) {
                    check = false;
                    break;
                } else {
                    if (s.top() == '[') {
                        s.pop();
                    } else {
                       check = false;
                        break; 
                    }
                }
            } 
        }
        
        if (check && s.empty()) {
            cout << "yes\n";
        } else {
            cout << "no\n";
        }
    }

    return 0;
}